contracts/oracle/BaseOraclePTPendle.sol

```diff
@@ -0,0 +1,92 @@
+
+/// @title BaseOraclePTPendle
+/// @author Angle Labs, Inc.
+/// @notice Base oracle implementation for PT tokens on Pendle
+abstract contract BaseOraclePTPendle {
+    uint256 public constant BASE_18 = 1 ether;
+    uint256 public constant YEAR = 365 days;
+    // @notice The maximum implied rate for the underlying asset, if set well it allows to have a lower bound on the PT token price
+    uint256 public maxImpliedRate;
+    // @notice The duration of the TWAP used to calculate the PT price
+    uint32 public twapDuration;
+
+    /*//////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+                                                        EVENTS                                                      
+    //////////////////////////////////////////////////////////////////////////////////////////////////////////////////*/
+
+    event MaxImpliedRateUpdated(uint256 _maxImpliedRate);
+    event TwapPTDurationUpdated(uint256 _twapDuration);
+
+    constructor(uint256 _maxImpliedRate, uint32 _twapDuration) {
+        maxImpliedRate = _maxImpliedRate;
+        twapDuration = _twapDuration;
+    }
+
+    modifier onlyGovernorOrGuardian() {
+        _onlyGovernorOrGuardian();
+        _;
+    }
+
+    function _getQuoteAmount() internal view virtual returns (uint256) {
+        uint256 economicalLowerBound = _economicalPTLowerBoundPrice();
+        uint256 pendlePrice = _pendlePTPrice();
+        uint256 minPrice = economicalLowerBound > pendlePrice ? pendlePrice : economicalLowerBound;
+        uint256 quote = (_detectHackRatio() * minPrice) / BASE_18;
+        return quote;
+    }
+
+    function setMaxImpliedRate(uint256 _maxImpliedRate) external onlyGovernorOrGuardian {
+        maxImpliedRate = _maxImpliedRate;
+        emit MaxImpliedRateUpdated(_maxImpliedRate);
+    }
+
+    function setTwapDuration(uint32 _twapDuration) external onlyGovernorOrGuardian {
+        twapDuration = _twapDuration;
+        emit TwapPTDurationUpdated(_twapDuration);
+    }
+
+    /*//////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+                                                       INTERNAL                                                     
+    //////////////////////////////////////////////////////////////////////////////////////////////////////////////////*/
+
+    function _economicalPTLowerBoundPrice() internal view returns (uint256) {
+        uint256 exp = block.timestamp > maturity() ? 0 : maturity() - block.timestamp;
+        if (exp == 0) return BASE_18;
+
+        UD60x18 denominator = UNIT.add(ud(maxImpliedRate)).pow(ud(exp).div(ud(YEAR)));
+        uint256 lowerBound = UNIT.div(denominator).unwrap();
+        return lowerBound;
+    }
+
+    function _pendlePTPrice() internal view returns (uint256) {
+        return PendlePtOracleLib.getPtToAssetRate(IPMarket(market()), twapDuration);
```

Comment:
Note that this oracle can only works for certain SY as written [below](https://docs.pendle.finance/Developers/Integration/HowToIntegratePtAndLpOracle#fourth-get-the-price):
<img width="947" alt="image" src="https://github.com/AngleProtocol/borrow-contracts/assets/44097430/8f1f4cff-71ea-45f8-9c6a-1f94693e79af">

Reply:
Very good remark, I'll make sure if we deploy a new oracle that it can also be used - review from Pendle And you (possibly)